play-json-tools/src/main/scala/com/evolutiongaming/util/JsonFormats.scala

```diff
@@ -234,10 +235,43 @@ object JsonFormats {
     }
 
     def writes(t: String, json: JsObject = Json.obj()): JsObject = {
+      if (json.keys contains "type") logger warn s"Inner JSON for '$t' subtype already contains 'type' field"
       Json.obj("type" -> t) ++ json
     }
   }
 
+  trait FoldedTypeFormat[T] extends OFormat[T] {
```

Comment:
I would advice to stop using `trait` but rather create method `foldedTypeFormat` which takes overridables as arguments.

Reply:
could you also cover this with unit test ?